src/api/ml/CMakeLists.txt

```diff
@@ -0,0 +1,320 @@
+          "-I" "${z3ml_bin}"
+          "${z3ml_bin}/z3ml.cmxa"
+  DEPENDS
+          libz3_z3ml
+          ${z3ml_bin}/z3native_stubs.o
+          ${z3ml_bin}/z3enums.cmo
+          ${z3ml_bin}/z3native.cmo
+          ${z3ml_bin}/z3.cmo
+          ${z3ml_bin}/z3enums.cmx
+          ${z3ml_bin}/z3native.cmx
+          ${z3ml_bin}/z3.cmx
+  COMMENT "Building z3ml.{cma,cmxa,cmxs}, dllz3ml.so, and libz3ml.a"
+  VERBATIM)
+
+###############################################################################
+# Example
+###############################################################################
+ 
+add_custom_command(
+  OUTPUT  ${z3ml_bin}/ml_example.ml
+  COMMAND "${CMAKE_COMMAND}" "-E" 
+          "copy" "${PROJECT_SOURCE_DIR}/examples/ml/ml_example.ml" "${z3ml_bin}/ml_example.ml"
+  DEPENDS ${PROJECT_SOURCE_DIR}/examples/ml/ml_example.ml
+  COMMENT "Copying ml_example.ml to build area")
+
+execute_process(
+  COMMAND ${OCAMLFIND} query zarith
+  OUTPUT_VARIABLE ocaml_pkg_zarith_path
+  OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+#     "-I" "+threads"
+
+add_custom_command(
+  OUTPUT ${z3ml_bin}/ml_example${bc_ext}
+         ${z3ml_bin}/ml_example.bc.log
+  COMMAND "${OCAMLFIND}" "ocamlc"
+    "-o" "${z3ml_bin}/ml_example${bc_ext}"
+    "-package" "zarith"
+    "-I" "${z3ml_bin}"
+    "-dllpath" "${z3ml_bin}"
+    "-I" ${ocaml_stublibs_path}
+    "-dllpath" ${ocaml_stublibs_path}
+    "${ocaml_pkg_zarith_path}/zarith.cma"
+    "${z3ml_bin}/z3ml.cma"
+    "${z3ml_bin}/ml_example.ml"
+  COMMAND 
+    "ocamlrun" "${z3ml_bin}/ml_example${bc_ext}"
+    ">" "${z3ml_bin}/ml_example.bc.log"
+  DEPENDS
+    ${z3ml_bin}/z3ml.cma
+    ${z3ml_bin}/dllz3ml.so
+    ${z3ml_bin}/ml_example.ml
+  COMMENT "Testing build and run ml_example bytecode"
+  VERBATIM)
+
+#     "-I" "+threads"
+
+add_custom_command(
+  OUTPUT ${z3ml_bin}/ml_example${exe_ext}
+         ${z3ml_bin}/ml_example.log
+  COMMAND "${OCAMLFIND}" "ocamlopt"
+    "-o" "${z3ml_bin}/ml_example${exe_ext}"
+    "-package" "zarith"
+    "-linkpkg"
+    "-I" "${z3ml_bin}"
+    "${z3ml_bin}/z3ml.cmxa"
+    "${z3ml_bin}/ml_example.ml"
+  COMMAND "${z3ml_bin}/ml_example${exe_ext}"
+    ">" "${z3ml_bin}/ml_example.log"
+  DEPENDS
+    ${z3ml_bin}/z3ml.cmxa
+    ${z3ml_bin}/dllz3ml.so
+    ${z3ml_bin}/ml_example.ml
+  COMMENT "Testing build and run ml_example natively"
+  VERBATIM)
+
+add_custom_target(build_ocaml_example
+  ALL
+  DEPENDS
+    ${z3ml_bin}/ml_example${bc_ext}
+    ${z3ml_bin}/ml_example${exe_ext}
+)
+
+add_custom_target(build_z3_ocaml_bindings
+  ALL
+  DEPENDS
+    ${z3ml_bin}/z3ml.cma
+    ${z3ml_bin}/z3ml.cmxa
+    ${z3ml_bin}/z3ml.cmxs
+    ${z3ml_bin}/dllz3ml.so
+    ${z3ml_bin}/libz3ml.a
+    build_ocaml_example
+)
+
+###############################################################################
+# Install
+###############################################################################
+
+# Hacky: When the os is APPLE, a fix command will mutate `libz3.dylib` and `dlllibz3.so` inplace.
```

Comment:
This is not so out-of-date but I will update it when macOS's solution for both building and installation is done. In the end, I may still leave some comments for macOS because it's kind of unlike Linux.